pkg/apiserver/registry/inventory/virtualmachine/rest.go

```diff
@@ -197,3 +180,148 @@ func (r *REST) Watch(ctx context.Context, options *internalversion.ListOptions)
+func getSelectors(options *internalversion.ListOptions) (map[string]string, map[string]string, error) {
+	var validLabelSelectors, validFieldSelectors map[string]string
+
+	validLabelSelectors = make(map[string]string, numLabelSelectors)
+	validFieldSelectors = make(map[string]string, numFieldSelectors)
+	if err := getLabelSelectors(options, validLabelSelectors); err != nil {
+		return nil, nil, err
+	}
+	if err := getFieldSelectors(options, validFieldSelectors); err != nil {
+		return nil, nil, err
+	}
+	return validLabelSelectors, validFieldSelectors, nil
+}
+
+// getLabelSelectors extracts and populates the label selectors map from the list options.
+func getLabelSelectors(options *internalversion.ListOptions, selectors map[string]string) error {
+	if options != nil && options.LabelSelector != nil && options.LabelSelector.String() != "" {
+		labelSelectorStrings := strings.Split(options.LabelSelector.String(), ",")
+		for _, labelSelectorString := range labelSelectorStrings {
+			labelKeyAndValue := strings.Split(labelSelectorString, "=")
+			if labelKeyAndValue[0] == cloudAccountName {
+				selectors[cloudAccountName] = labelKeyAndValue[1]
+			} else if labelKeyAndValue[0] == cloudAccountNamespace {
+				selectors[cloudAccountNamespace] = labelKeyAndValue[1]
+			} else {
+				return errors.NewBadRequest(fmt.Sprintf("unsupported label selector, supported labels are: "+
+					"%v and %v", cloudAccountName, cloudAccountNamespace))
+			}
+		}
+	}
+	return nil
+}
+
+// getFieldSelectors extracts and populates the field selectors map from the list options.
+func getFieldSelectors(options *internalversion.ListOptions, selectors map[string]string) error {
+	if options != nil && options.FieldSelector != nil && options.FieldSelector.String() != "" {
+		fieldSelectorStrings := strings.Split(options.FieldSelector.String(), ",")
+		for _, fieldSelectorString := range fieldSelectorStrings {
+			fieldKeyAndValue := strings.Split(fieldSelectorString, "=")
+			if fieldKeyAndValue[0] == metaName {
+				selectors[metaName] = fieldKeyAndValue[1]
+			} else if fieldKeyAndValue[0] == metaNamespace {
+				selectors[metaNamespace] = fieldKeyAndValue[1]
+			} else if fieldKeyAndValue[0] == statusCloudId {
+				selectors[statusCloudId] = fieldKeyAndValue[1]
+			} else if fieldKeyAndValue[0] == statusCloudVpcId {
+				selectors[statusCloudVpcId] = fieldKeyAndValue[1]
+			} else if fieldKeyAndValue[0] == statusRegion {
+				selectors[statusRegion] = fieldKeyAndValue[1]
+			} else {
+				return errors.NewBadRequest(fmt.Sprintf("unsupported field selector, supported fields are: "+
+					"%v, %v, %v, %v and %v", metaName, metaNamespace, statusCloudId, statusCloudVpcId, statusRegion))
+			}
+		}
+	}
+	return nil
+}
+
+// matchLabel matches vm label with the provided selector and returns
+// true if a match is found.
+func matchLabel(vmLabels, selector map[string]string, name string) bool {
+	if selector[name] != "" {
+		value, ok := vmLabels[name]
+		if !ok {
```

Comment:
is this logically correct? second lookup will only work if first lookup is successful

Reply:
We want to check if the selector[] contains a value set for the given label.
If it is set then proceed further.